### ISSUE TITLE
Optionally disable writer starvation protection?

### ISSUE THREAD

**Issue by SimonSapin** [Amanieu/parking_lot#27]
From the docs of `RwLock`:

> This lock uses a task-fair locking policy which avoids both reader and writer starvation. This means that readers trying to acquire the lock will block even if the lock is unlocked when there are writers waiting to acquire the lock. Because of this, attempts to recursively acquire a read lock within a single thread may result in a deadlock.

I’m looking into building an abstraction on top of parking_lot to protect many objects with a single `RwLock<()>`. In the expected access patterns, I believe that starvation won’t be a problem but the deadlock with recursive reader access might be. Is it possible to configure `RwLock` to tweak this behavior?

CC @bholley

**Comment by Amanieu** [comment]
In theory, a design based which favors readers over writers (this is the default for pthread rwlocks) would allow recursive read locks to work without the possibility of deadlocks. However I am uncomfortable with making this the default since more rwlock workloads tend to be reader-heavy, in which case you really want to avoid writer starvation.

Maybe we can make it an option when the `RwLock` is constructed?

**Comment by SimonSapin** [comment]
Yes, I didn’t mean that this should be the default. Having some way to opt into this behavior could be useful. I’m now sure how, though.

A separate constructor method would work, but if the `RwLock` struct needs an extra `bool` field its size would grow by up to 8 bytes because of alignment, which doesn’t seem great. Using a different type would allow different behaviors without a runtime `bool` field, but doubles the API surface. `RwLock` could gain an extra type parameter with a default type, but IIRC default parameters don’t yet guide type inference as much as they should.

**Comment by arthurprs** [comment]
This is probably hard to implement efficiently in a single type, what about a new BiasedRwLock?

**Comment by Amanieu** [comment]
It could be done in a single type by just stealing a bit in the `RawRwLock`. I think this is possible, and could be exposed as an alternate constructor, but I'll need to double check the implementation.

**Comment by arthurprs** [comment]
Wouldn't you have to `and` and `branch` on it in every op? That might slow show on benchmarks and pollute the branch predictor cache.

**Comment by Amanieu** [comment]
The fast paths would remain mostly unchanged (you can ignore the extra bit in most cases), it's mostly the slow path that needs to change how it wakes up threads.

I'll give it a try and see if it impacts performance.

**Comment by Amanieu** [comment]
So I'm considering something along these lines:
```rust
trait RwLockPolicy {
    fn to_dynamic(&self) -> RwLockPolicyDynamic;
}

struct RwLockPolicyPreferReaders;
struct RwLockPolicyPreferWriters;
struct RwLockPolicyPhaseFair;
struct RwLockPolicyTaskFair;
enum RwLockPolicyDynamic {
    PreferReaders,
    PreferWriters,
    PhaseFair,
    TaskFair
}

struct RwLock<T, P: RwLockPolicy = RwLockPolicyPreferWriters>;
impl RwLock<T, P: RwLockPolicy = RwLockPolicyPreferWriters> {
    pub fn new(val: T) -> RwLock<T> {}
    pub fn with_policy(val: T, policy: P) -> RwLock<T, P> {}
}
```

This allows you to choose a policy either statically or dynamically. Only the reader-prefering policy will allow recursive read locks. Thoughts?

**Comment by arthurprs** [comment]
I think these should be statically only, along some handy type aliases like '`type RwLockReadBiased: ...` and so on.

**Comment by CryZe** [comment]
Welp, I just massively ran into this issue last night as well. Took me like 5 hours to figure out why my code is deadlocking, so I would appreciate some solution for this ^^

It's not too bad though, as I can at least somewhat avoid recursive reads.

**Comment by Amanieu** [comment]
After trying to implement this for a bit, I think that this is too complicated in practice. I'm currently tending towards just adding these methods:
```rust
fn read_recursive(&self)
fn try_read_recursive(&self)
fn try_read_recursive_for(&self, timeout: Duration)
fn try_read_recursive_until(&self, timeout: Instant)
```

These would work just like the existing read locks, except that they won't block on recursive read locks, at the cost of potentially incurring writer starvation (readers keep grabbing the lock without giving writers a chance to grab it).

Would this be good enough for your use cases?

**Comment by arthurprs** [comment]
Wouldn't it be better as a separated type altogether, like, RecursiveRwLock?

**Comment by Amanieu** [comment]
The problem with a separate type is that it introduces a lot of complexity... (need to add an extra type parameter to `RwLock`, create typedefs, etc)

**Comment by arthurprs** [comment]
Probably not worth it then :smile:

**Comment by Amanieu** [comment]
Implemented in 0.4.4.
